### DISCUSSION TITLE
Best practice of sampling?

### DISCUSSION THREAD

**Discussion by 1393467261** [apache/skywalking#8110] [Q&A] [answered]
Good afternoon gentlemen. I am wandering about the best practice of sampling. Maybe we can discuss the `forceSampleErrorSegment`,`slowDBAccessThreshold`,`slowTraceSegmentThreshold`(those which will cause incomplete trace) and the proper sample rate during the coffee time.

**Comment by wu-sheng** [comment, marked as answer]
The key point of sampling is reducing the size of data set, and keep as much information as possible with reasonable resources used. Any of this is designed to keep complete distributed traces. In theory, only 100% sampling could keep that.

**Reply by 1393467261** [reply]
I see, `as much information as possible` good idea. And any suggestion for the sample rate? By the way my current rate is 100%.

**Reply by wu-sheng** [reply]
That depends on your budget, this is never being a tech decision. I am afraid. 100% as default are because, 1) SkyWalking provides this capability rather than others(which prefer sampling always), 2) This is friendly for new users to quick start.

**Reply by 1393467261** [reply]
Got it, thank you.

**Reply by 1393467261** [reply]
And one more question: the ignored traces will be excluded from the metric calculation or not?

**Reply by wu-sheng** [reply]
Backend sampling doesn't have side effect for metrics. 
But if sampling happens in agent side, metrics lost accordingly.

**Comment by xiaozaiyuji** [comment]
Number 1:
//The number of sampled traces per 3 seconds
//Negative or zero means off, by default
agent.sample_n_per_3_secs=${SW_AGENT_SAMPLE:120}
Number 2:
receiver-trace:
  selector: ${SW_RECEIVER_TRACE:default}
  default:
      sampleRate: ${SW_TRACE_SAMPLE_RATE:1000}
Number 3：
default:
  // Default sampling rate that replaces the 'agent-analyzer.default.sampleRate'
  // The sample rate precision is 1/10000. 10000 means 100% sample in default.
  rate: 10000
  // Default trace latency time that replaces the 'agent-analyzer.default.slowTraceSegmentThreshold'
 //Setting this threshold about the latency would make the slow trace segments sampled if they cost more time, even the sampling mechanism activated. The default value is `-1`, which means would not sample slow traces. Unit, millisecond.
  duration: -1
I was confusion about this three config param. Can I get some help about the use tips about the three config param. I do not know the relations  between the three config param about the data is cut to store in elasticsearch if the storage is elasticsearch.

**Reply by xiaozaiyuji** [reply]
Can you help me? @wu-sheng  Thanks very much.

**Reply by wu-sheng** [reply]
If you have read the documents of these configurations, you should have known.
Backend and agent sampling mechanisms are irrelevant. There is no point to discuss them together.
